### PR TITLE
[CAL-770] add new integration architecture revamp

### PR DESCRIPTION
**Pull request by edanfesi**
## What does this PR do?

Develop the new integrations architecture revamp for the calendars

Fixes # (issue)

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

## How should this be tested?

- [x] Run existing test suits because the project has to run normally

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code and corrected any misspellings
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: tsconfig.tsbuildinfo [hunk 1]
```diff
@@ -0,0 +1 @@
...c08ea2e05e34ff7b6ba5b79806f625d8b483ec25bd2f8e1c2f013a0","e57f0b3ed5707bd45fa3a0d54faef6c1543e6374e6f516141dabb753dde6dff3","c86fea295c21ea01c93410eba2ec6e4f918b97d0c3bf9f1bb1960eabe417e7eb","05d41b3e7789381ff4d7f06d8739bf54cc8e75b835cb28f22e59c1d212e48ff3","ff8c485f6d39157c29e9196f99b0e69b13508b84bde6b51b01f8f3f7cb35f4b8","94169a40e1ac690c161c8e61b388d298ab202c9b95a885532d2e54686e24adb3","f63be9b46a22ee5894316cf71a4ba7581809dd98cf046109060a1214ee9e2977","dd3cc41b5764c9435b7cae3cc830be4ee6071f41a607188e43aa1edeba4fbb3e","b2dbb9485701a1d8250d9a35b74afd41b9a403c32484ed40ed195e8aa369ae70","5aa7565991c306061181bd0148c458bcce3472d912e2af6a98a0a54904cd84fc","9629e70ae80485928a562adb978890c53c7be47c3b3624dbb82641e1da48fd2f","c33d86e1d4753d035c4ea8d0fdb2377043bc894e4227be3ceabc8e6a5411ab2e","f9ec74382c95cbc85804daf0e9dabed56511a6dfb72f8a2868aa46a0b9b5eafc","be32c0a0576265a4dee467f328c5945805a832e6268d312ed768cae1f2666fa6","f03ad8ed9d468e4bd52cbc20888bc72df27aabab4b5d57f917c5a7de0e9a9bee","e4ae60dfe4f3b266e1880c565e6830349bd67502aaed04d4196c14f006610e4f","4dc6a62b37bbe4af66ef387690b6493c984eee96d0e5e9979f17ecdc098f520e","90878ed33999d4ff8da72bd2ca3efb1cde76d81940767adc8c229a70eb9332b2","d72
```
Comment (zomars) [inline tsconfig.tsbuildinfo]: This file should not be committed @edanfesi Can we removed it?

**Review by zomars — CHANGES_REQUESTED** [review]
Looking good! Let's just address some comments and make sure tests are passing. Also, we should add basic integration tests for each calendar. What do you think?

**Review by zomars — APPROVED** [review]

File: playwright/integrations/calendar/integrations-google-calendar.test.ts [hunk 2]
```diff
@@ -0,0 +1,88 @@
+import { expect, test } from "@playwright/test";
```
Comment (zomars) [inline playwright/integrations/calendar/integrations-google-calendar.test.ts:1]: Let's remove this test for this PR and add it in a follow up @edanfesi

**Review by zomars — CHANGES_REQUESTED** [review]

File: lib/integrations/calendar/services/CalDavCalendarService.ts [hunk 3]
```diff
@@ -0,0 +1,10 @@
+import { Credential } from "@prisma/client";
+
+import { CALENDAR_INTEGRATIONS_TYPES } from "../constants/generals";
+import CalendarService from "./BaseCalendarService";
+
+export default class CalDavCalendarService extends CalendarService {
+  constructor(credential: Credential) {
```
Comment (emrysal) [inline lib/integrations/calendar/services/CalDavCalendarService.ts:7]: Just general observation that using inheritance to build a factory pattern is a way to do it I suppose, but it results in a lot of bloat.

**Review by emrysal — COMMENTED** [review]
Overall the main improvement here is the being made redundant of the calendarClient 👍  but some of the decisions here I highlighted as excessive; some other feedback as well.

File: lib/integrations/calendar/services/BaseCalendarService.ts [hunk 4]
```diff
@@ -18,73 +18,62 @@ import { getLocation, getRichDescription } from "@lib/CalEventParser";
...
 
-import { CalendarEvent, IntegrationCalendar } from "./calendarClient";
+import { TIMEZONE_FORMAT } from "../constants/formats";
+import { CALDAV_CALENDAR_TYPE } from "../constants/generals";
+import { CalendarEventType, EventBusyDate, NewCalendarEventType } from "../constants/types";
+import { Calendar, CalendarEvent, IntegrationCalendar } from "../interfaces/Calendar";
+import { convertDate, getAttendees, getDuration } from "../utils/CalendarUtils";
 
-dayjs.extend(utc);
+const { CALENDSO_ENCRYPTION_KEY = "" } = process.env;
```
Comment (zomars) [inline lib/integrations/calendar/services/BaseCalendarService.ts:27]: We shouldn't destructure env variables according to [Next.js docs](https://nextjs.org/docs/api-reference/next.config.js/environment-variables):

> Next.js will replace process.env.customKey with 'my-value' at build time. Trying to destructure process.env variables won't work due to the nature of webpack DefinePlugin.

```suggestion
const CALENDSO_ENCRYPTION_KEY = process.env.CALENDSO_ENCRYPTION_KEY || "";
```

**Review by emrysal — APPROVED** [review]
Happy with the recent changes, it's a good step forward :)

File: lib/integrations/calendar/services/GoogleCalendarService.ts [hunk 5]
```diff
@@ -0,0 +1,324 @@
...
+import { getLocation, getRichDescription } from "@lib/CalEventParser";
+import { CALENDAR_INTEGRATIONS_TYPES } from "@lib/integrations/calendar/constants/generals";
+import logger from "@lib/logger";
+import prisma from "@lib/prisma";
+
+import { EventBusyDate, NewCalendarEventType } from "../constants/types";
+import { Calendar, CalendarEvent, IntegrationCalendar } from "../interfaces/Calendar";
+import CalendarService from "./BaseCalendarService";
+
+const { GOOGLE_API_CREDENTIALS = "" } = process.env;
```
Comment (zomars) [inline lib/integrations/calendar/services/GoogleCalendarService.ts:14]: Same as above

```suggestion
const GOOGLE_API_CREDENTIALS = process.env.GOOGLE_API_CREDENTIALS || "";
```

File: lib/integrations/calendar/services/CalDavCalendarService.ts (hunk 3 above)
Comment (zomars) [reply lib/integrations/calendar/services/CalDavCalendarService.ts:7]: I find it very neat being able to reuse as much logic as possible. How is this bloated?

File: lib/integrations/calendar/services/Office365CalendarService.ts [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+import { handleErrorsJson, handleErrorsRaw } from "@lib/errors";
+import { CALENDAR_INTEGRATIONS_TYPES } from "@lib/integrations/calendar/constants/generals";
+import logger from "@lib/logger";
+import prisma from "@lib/prisma";
+
+import { BatchResponse, EventBusyDate, NewCalendarEventType } from "../constants/types";
+import { Calendar, CalendarEvent, IntegrationCalendar } from "../interfaces/Calendar";
+import { BufferedBusyTime, O365AuthCredentials } from "../interfaces/Office365Calendar";
+
+const { MS_GRAPH_CLIENT_ID = "", MS_GRAPH_CLIENT_SECRET = "" } = process.env;
```
Comment (zomars) [inline lib/integrations/calendar/services/Office365CalendarService.ts:14]: I also prefer your style, but this is a webpack plugin limitation. 😞 
```suggestion
const MS_GRAPH_CLIENT_ID = process.env.MS_GRAPH_CLIENT_ID || "";
const MS_GRAPH_CLIENT_SECRET = process.env.MS_GRAPH_CLIENT_SECRET || "";
```

File: tsconfig.tsbuildinfo [hunk 7]
```diff
@@ -0,0 +1 @@
...5137af10fed17807a20df861ab8","30884e6c4e805be89fc1178ce92ff95adff50e5330246a525db72b8501cdd698","2a8fe3e7bedb25f587783d0a02b027f4bc9b65ce1f93e4d034a2826582cc1c17","bacf61766d6b2a9662d48b0adb6ffd9cf116a179862c8b6698c7b98bd3e0cf19","cf4dc15ca9dc6c0995dd2a9264e5ec37d09d9d551c85f395034e812abdf60a99","e74babac0be57a8e13eb69efb67286c6e954625978c67449e679267180ded327","4cb49e79595c6413fcb01af55a8a574705bf385bd2ec5cf8b777778952e2914a","d6b44382b2670f38c8473e7c16b6e8a9bfa546b396b920afc4c53410eeb22abf","3b5c6f451b7ad87e3fcd2008d3a6cb69bd33803e541e9c0fe35754201389158f","548ef4a7ba09bdd5a03e3be32ecfd4efac910d86691444f80f56d95bd65d6d9d","a2d163941b598e23edc92623f5538fb8dfa84014009af271f1b2b5d7c29fdccb","48a019b68514f062d9da21319e2258088a75c6c0175031748b036ed9bf464fd7","7d3f8373b871b9570f85f1f3d598f186c547fbc20bb47f606c7055d73acee9fd","a01950b95ce3146810d9390e8561329610a47e127d5f9518c89ec18c25179bfe","0207f87632dd8a801dcd0eaa423a4d5bb35ea39ccc60b8c21c475429887679f9","59e9defd647d02a00ba6f6fd737535664703ff74b95a94868c020edd8027b79c","75ef655e0a83bed5c95349e78f66f29375d9c034ba2ffa43e7514d5cf0bbc627","d14dbf3a0c3d3265e109c96495a16999dd64b4a23bddca16111d15833c5fb271","921274e649ce4eafc40a63f0f155bd5
```
Comment (zomars) [inline tsconfig.tsbuildinfo]: This file shouldn't be committed.

**Review by zomars — COMMENTED** [review]
Awesome work @edanfesi We're almost there. Just some minor details left.

File: lib/integrations/calendar/services/BaseCalendarService.ts (hunk 4 above)
Comment (zomars) [reply lib/integrations/calendar/services/BaseCalendarService.ts:27]: Also it's a [security concern](https://nextjs.org/docs/basic-features/environment-variables): 

> Note: In order to keep server-only secrets safe, Next.js replaces process.env.* with the correct values at build time. This means that process.env is not a standard JavaScript object, so you’re not able to use object destructuring. Environment variables must be referenced as e.g. process.env.PUBLISHABLE_KEY, not const { PUBLISHABLE_KEY } = process.env.

**Review by zomars — APPROVED** [review]
all gucci, nice work! 🙌🏽

**Pull request merged**